toughday/src/main/java/com/adobe/qe/toughday/internal/core/distributedtd/redistribution/TaskBalancer.java

```diff
@@ -0,0 +1,287 @@
+                        if (!response.equals(Agent.Status.RUNNING.toString())) {
+                            ready = false;
+                            break;
+                        }
+                    } catch (IOException e) {
+                        LOG.warn("Could not confirm that agent " + agentIpAddress + " is ready for processing " +
+                                "redistribution instructions") ;
+                    }
+                }
+            }
+
+            try {
+                Thread.sleep(1000);
+            } catch (InterruptedException e) {
+                // skip this and continue waiting for all agents to be ready for redistribution
+            }
+        }
+    }
+
+    private void sendInstructionsToExistingAgents(Map<String, Phase> phases, Queue<String> activeAgents) {
+        ObjectMapper mapper = new ObjectMapper();
+
+        // send instructions to agents that are already running TD
+        activeAgents.stream()
+            .filter(ipAddress -> !this.inactiveAgents.contains(ipAddress)) // filter agents that become inactive
+            .forEach(ipAddress -> {
+                String agentURI = Agent.getRebalancePath(ipAddress);
+                TestSuite testSuite = phases.get(ipAddress).getTestSuite();
+                RunMode runMode = phases.get(ipAddress).getRunMode();
+
+                // build redistribution instructions
+                Map<String, Long> testSuiteProperties = getTestSuitePropertiesToRedistribute(testSuite);
+                Map<String, String> runModeProperties = runMode.getRunModeBalancer()
+                        .getRunModePropertiesToRedistribute(runMode.getClass(), runMode);
+                RedistributionInstructions redistributionInstructions =
+                        new RedistributionInstructions(testSuiteProperties, runModeProperties);
+                try {
+                    String message = mapper.writeValueAsString(redistributionInstructions);
+                    LOG.info("[Redistribution] Sending " + message + " to agent " + ipAddress);
+
+                    HttpResponse response = this.httpUtils.sendHttpRequest(HttpUtils.POST_METHOD, message,
+                            agentURI, HTTP_REQUEST_RETRIES);
+                    if (response == null) {
+                        /* redistribution will be automatically triggered as soon as this agent fails to respond to
+                         * heartbeat request
+                         */
+                        LOG.warn("Redistribution instructions could not be sent to agent " + ipAddress + ".");
+                    }
+
+                } catch (JsonProcessingException e) {
+                    LOG.error("Unexpected exception while sending redistribution instructions", e);
+                    LOG.warn("Agent " + ipAddress + " will continue to run with the configuration it had before " +
+                            "the process of redistribution was triggered.");
```

Comment:
The properties of the current Phase of the execution will no longer be respected(for instance, the desired load might not be achieved). The next phase will execute properly, so I believe it is not necessary to kill the entire distributed execution because of this.

This exception should never be thrown as long as the RedistributionInstructions class has a format that allows the ObjectMapper to dump it in JSON format.